### DISCUSSION TITLE
How to copy a Record object?

### DISCUSSION THREAD

**Discussion by kentaroi** [groue/GRDB.swift#1494] [Q&A] [answered]
`Record.copy()` was removed in https://github.com/groue/GRDB.swift/commit/a15f731042767a5626ee69a9a5b70f2e8e4a1712
How to copy a Record object?
I cannot mimic [the removed `Record.copy()`](https://github.com/groue/GRDB.swift/blob/3000cd3c7f3577827b2185944473ea25ae0f0456/GRDB/Record/Record.swift#L152-L156) since `Row(self)` seems to be internal.
I can use `JSONEncoder` and `JSONDecoder` or manually copy properties, but cannot reset `referenceRow` variable or call `resetDatabaseChanges()` since they are internal or private.

Thanks in advance!

**Comment by groue** [comment]
Hello @kentaroi,

It was removed when `FetchableRecord.init(row:)` has become a throwing initializer.

We could reintroduce `copy()`, but it would have to be a throwing method. And... you're the first user asking for it, so it does not look like a frequent use case.

May I ask if you have considered some alternatives?

- Use structs instead of `Record` subclasses? Structs don't need the `copy` method:

    ```swift
    let p1: Player = ...
    let p2 = p1 // A copy, freely given by the language
    ```

- ~Write your own version of `copy()`? It would be much more efficient than a serialisation-deserialization from `Row` or from JSON.~ Oh no you can't because `referenceRow` is not public.

- If you use methods that depend on `referenceRow`, which gives `Record` the unique ability to compare against itself, did you have a look at the features available for non-`Record` records, in [Record Comparison](https://github.com/groue/GRDB.swift/blob/master/README.md#record-comparison)?

On the last bullet, I'm constantly using [`updateChanges(_:modify:)`](https://swiftpackageindex.com/groue/grdb.swift/documentation/grdb/record/updatechanges(_:onconflict:modify:)):

```swift
struct Player { ... }

// Updates the database if and only if `player` was modified:
var player: Player = ...
player.updateChanges(db) {
    $0.score = 1000
}
print(player.score) // 1000
```

This method alone has completely removed my own needs for self-comparing records.

**Comment by kentaroi** [comment]
Thank you very much for your quick response, @groue!!
And, thanks for the suggestions.

I’ve implemented some records both ways, with `Record` classes and with structs with protocols of GRDB, and I think I understand the pros of using structs.

However, as you know, when structs are complicated, they can be cumbersome.
For example:
```Swift
struct A {
    var name: String = “A” {
        didSet {
            print(“A.name is updated”)
        }
    }
}

struct B {
    var a: A = A() {
        didSet {
            print(“B.a is updated”)
        }
    }
}

struct C {
    var b: B = B() {
        print(“C.b is updated”)
    }
}
var c = C()
c.b.a.name = “a”
```

If you execute the above code, you will see the following output:
```
A.name is updated
B.a is updated
C.b is updated
```
instead of `A.name is updated`.
I use associations and my structs are a bit complicated. And to make things work correctly, in the property observers(`didSet`), I have to check if it actually changed the value of the property or a property or a descendant property of the value is changed.

Therefore, I think `Record` class is helpful sometimes.

I don't think structs and protocols of GRDB have any problems, but sometimes I'd like to use classes for some purposes in Swift in general especially when handling hierarchical objects.

Is it possible to make `resetDatabaseChanges()` accessible or create a way to make `hasDatabaseChanges` be false, or reintroduce `copy()` as a throwing method as you wrote?

**Reply by groue** [reply]
> I don't think structs and protocols of GRDB have any problems, but sometimes I'd like to use classes for some purposes in Swift in general especially when handling hierarchical objects.

That's a valid claim 👍 

> Is it possible to make `resetDatabaseChanges()` accessible or create a way to make `hasDatabaseChanges` be false, or reintroduce `copy()` as a throwing method as you wrote?

You ask your question at the perfect moment, because GRDB 7 is in the making. Shipping a new major version is the opportunity to review existing apis, introduce breaking changes, and enable new features.

In the next features that I wish would be enabled by GRDB 7 is the support for [JSONB](https://www.sqlite.org/json1.html#jsonbx). I did not start the analysis of the necessary changes yet, but I will probably have to change the conversion of records to and from `Row`. In the process, _I'm not sure if `Row(record)` will still be implementable._ This would remove the necessary foundation for `copy()`.

It's too early for me to have a clear opinion on this. It's just a word of caution from my little finger.

On the other hand, it should be possible to implement `resetDatabaseChanges()`, even if `Row(record)` no longer exists.

In the end, I'm sorry to give you some work, but would you please explain your use cases, the way you want to use `copy()` or `resetDatabaseChanges()`? This would help figuring out the best api(s) that allow you to program your application, while preserving the desired low-level changes that will enable the support for JSONB?

**Reply by kentaroi** [reply]
Thank you very much for considering it, @groue.

Below is my use case. To make it easier to visualize, I write to the point where it may not be necessary. Sorry for the verbosity.

I am building a kind of timer app.
There are models like TimerTemplate, Routine, and Favorite.
Routine `hasMany` TimerTemplates.
Favorite `belongsTo` both TimerTemplate and Routine, but in fact a Favorite object can belong to only one of them (this is enforced by the application code).

```Swift
class TimerTemplate: Record, Codable {
    static let favorite = hasOne(Favorite.self)
    static let routine  = belongsTo(Routine.self)
}

class Routine: Record, Codable {
    static let favorite       = hasOne(Favorite.self)
    static let timerTemplates = hasMany(TimerTemplate.self)
}

class RoutineInfo: FetchableRecord, Codable {
    var routine: Routine
    var timerTemplates: [TimerTemplate]
}

class Favorite: Record, Codable {
    static let timerTemplate = belongsTo(TimerTemplate.self)
    static let routine       = belongsTo(Routine.self)
}

class FavoriteInfo: FetchableRecord, Codable {
    var favorite: Favorite
    var timerTemplate: TimerTemplate?
    var routineInfo: RoutineInfo?
}

class MiniTimer: Codable {
    var favoriteInfo: FavoriteInfo? {
        didSet {
            guard let favoriteInfo = favoriteInfo else {
                timerTemplate = TimerTemplate()
                return
            }
            if let timerTemplate = favoriteInfo.timerTemplate {
                self.timerTemplate = timerTemplate
            } else {
                routineIndex = 0
            }
        }
    }

    var routineIndex: Int = 0 {
        didSet {
            guard let timerTemplates = favoriteInfo?.routineInfo?.timerTemplates else { return }
            timerTemplate = timerTemplates[routineIndex]
        }
    }

    var timerTemplate: TimerTemplate = TimerTemplate() {
        timeRemaining = timerTemplate.duration
        delegate?.miniTimerUpdatedTitle(self)
        delegate?.miniTimerUpdatedDuration(self)
    }

    var title: String? {
        get { timerTemplate.title }
        set(newTitle) {
            timerTemplate.title = newTitle
            delegate?.miniTimerUpdatedTitle(self)
        }
    }

    var duration: TimeInterval {
        get { timerTemplate.duration }
        set(newDuration) {
            timerTemplate.duration = newDuration
            delegate?.miniTimerUpdatedDuration(self)
        }
    }
}

```

When a user taps a FavoriteInfo cell in a table view displaying FavoriteInfos, the FavoriteInfo object is set to a MiniTimer object.
The user can manipulate the MiniTimer object.
For example, the user can change the `duration` of the MiniTimer, which will be reflected to the `duration` of the `timerTemplate` of the MiniTimer, which will be reflected in a TimerTemplate object in the `favoriteInfo` of the MiniTimer (because they are the same TimerTemplate object), but actually this `favoriteInfo` object is the same instance as a FavoriteInfo object in the the data source of the above table view, which is not what I want.

To prevent this, I need to deep copy the FavoriteInfo object before setting it to the MiniTimer object.

This way, the user can tweak `favoriteInfo` in the MiniTimer object without affecting to the data source of the table view, and can save the change if he/she want.

**Reply by groue** [reply]
I try to resist repeating that structs are the baked-in-the-language solution to this 😅

I can see how `copy()` would help in your scenario:

1. Fetch `Favorite`/`TimerTemplate`/`RoutineInfo` instances from the database.
2. Initialize `MiniTimer` instances containing copies of fetched values.
3. User can tweak those independent copies. Since those copies remember their reference row, you can query their `hasDatabaseChanges` property for any good purpose (conditionally display an "Apply" button, or optimize database accesses).

But you also said "Is it possible to make `resetDatabaseChanges()` accessible or create a way to make `hasDatabaseChanges` be false?". It looks like you see a way to implement your desired features with `resetDatabaseChanges()`, instead of `copy()`. I'm not sure I see how you'd do that?

**Reply by kentaroi** [reply]
I apologize if my previous comment was misleading.

>  For example, the user can change the duration of the MiniTimer, which will be reflected to the duration of the timerTemplate of the MiniTimer, which will be reflected in a TimerTemplate object in the favoriteInfo of the MiniTimer (because they are the same TimerTemplate object), 

This is what I want and I intentionally coded it that way.

> , but actually this favoriteInfo object is the same instance as a FavoriteInfo object in the the data source of the above table view

But, this is not what I want.

When the user tapped a cell in the table view, I would like to deep copy the favoriteInfo and set it to a MiniTimer object as follows:
```Swift
extension FavoritesViewController: UITableViewDelegate {
    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        guard let favoriteInfo = try? favoriteInfos[indexPath.row].copy() else { return }
        TimerService.shared.miniTimer.favoriteInfo = favoriteInfo
    }
}
```

> I try to resist repeating that structs are the baked-in-the-language solution to this 😅

You might be right. I implemented it in both ways, but I prefer the code using the `Record` class. Maybe it's because my understanding of structs is shallow.
Here is my MiniTimer code using structs for models:
```Swift
class MiniTimer: Codable {
    var favoriteInfo: FavoriteInfo? {
        didSet {
            guard
                (favoriteInfo != nil && oldValue == nil) ||
                (favoriteInfo == nil && oldValue != nil) ||
                (favoriteInfo != nil && oldValue != nil && favoriteInfo!.favorite.id != oldValue!.favorite.id)
            else {
                return
            }

            referenceFavoriteInfo = favoriteInfo
            guard let favoriteInfo = favoriteInfo else {
                _timerTemplate = TimerTemplate()
                return
            }
            routineIndex = 0
        }
    }

    var referenceFavoriteInfo: FavoriteInfo?

    var routineIndex: Int = 0 {
        didSet {
            timeRemaining = duration
            delegate?.miniTimerUpdatedTitle(self)
            delegate?.miniTimerUpdatedDuration(self)
        }
    }

    var _timerTemplate: TimerTemplate = TimerTemplate() {
        didSet {
            guard _timerTemplate.id != oldValue.id else { return }
            timeRemaining = _timerTemplate.duration
            delegate?.miniTimerUpdatedTitle(self)
            delegate?.miniTimerUpdatedDuration(self)
        }
    }

    var timerTemplate: TimerTemplate {
        get {
            guard let favoriteInfo = favoriteInfo else { return _timerTemplate }
            if let timerTemplate = favoriteInfo.timerTemplate {
                return timerTemplate
            } else {
                return favoriteInfo.routineInfo!.timerTemplates[routineIndex]
            }
        }
        set(newTimerTemplate) {
            if favoriteInfo != nil {
                if favoriteInfo?.timerTemplate != nil {
                    favoriteInfo?.timerTemplate = newTimerTemplate
                } else {
                    favoriteInfo?.routineInfo?.timerTemplates[routineIndex] = newTimerTemplate
                }
            } else {
                _timerTemplate = newTimerTemplate
            }
        }
    }

    var title: String? {
        get {
            return timerTemplate.title
        }
        set(newTitle) {
            timerTemplate.title = newTitle
            delegate?.miniTimerUpdatedTitle(self)
        }
    }

    var duration: TimeInterval {
        get { timerTemplate.duration }
        set(newDuration) {
            timerTemplate.duration = newDuration
            delegate?.miniTimerUpdatedDuration(self)
        }
    }
}
```


> But you also said "Is it possible to make resetDatabaseChanges() accessible or create a way to make hasDatabaseChanges be false?". It looks like you see a way to implement your desired features with resetDatabaseChanges(), instead of copy(). I'm not sure I see how you'd do that?

Here is a possible solution that I came up with:
```Swift
protocol CodableRecord: Record, Codable {
    func copy() throws -> Self
}

extension CodableRecord {
    func copy() throws -> Self {
        let encoder = JSONEncoder()
        let jsonData = try encoder.encode(self)

        let decoder = JSONDecoder()
        let clone = try decoder.decode(Self.self, from: jsonData)
        try clone.resetDatabaseChanges()
        return clone
    }
}
```

And, in my use case:
```Swift
extension Favorite: CodableRecord { }
extension TimerTemplate: CodableRecord { }

extension FavoriteInfo {
    func copy() throws -> FavoriteInfo {
        let favorite = try favorite.copy()
        if let timerTemplate = timerTemplate {
            return FavoriteInfo(favorite: favorite,
                                timerTemplate: try timerTemplate.copy(),
                                routineInfo: nil)
        } else {
            return FavoriteInfo(favorite: favorite,
                                timerTemplate: nil,
                                routineInfo: try routineInfo!.copy())
        }
    }
}
```

**Reply by groue** [reply]
> Here is a possible solution that I came up with:

I see. Yes, the record copied this way would be "clean".

OK. One last question: are you currently blocked? Or are you still using a version of GRDB that provides `copy()`?

**Reply by kentaroi** [reply]
> OK. One last question: are you currently blocked? Or are you still using a version of GRDB that provides copy()?

Yes, I am currently blocked.

**Reply by groue** [reply, marked as answer]
Ouch. I can't tell when the next version of GRDB will ship that will provide `copy` or any other API that fulfill your needs.

So, here are a few possible recommendations that will unblock you:

1. Reconsider structs. Maybe they'll click in your mind? Expected benefits: you'll never have such a problem in the future, and you won't have to wait for the next release. ⭐ 
2. Fork GRDB, make `resetDatabaseChanges()` public, and wait for the next release. 🤓 
3. Instead of copying the records, re-fetch them from the database? This looks like a very lousy workaround so I hope you won't pick this one 😅

**Reply by kentaroi** [reply]
Thank you for the suggestions!

I'm still grappling with how structs can streamline my code. If you have a spare moment, any pointers would be greatly appreciated.

I confess, the MiniTimer class code using model structs isn't quite hitting the mark for me right now. If I don't have an "aha!" moment within a day or two, I might consider forking GRDB.

Regardless, thank you again for your valuable guidance!
